### PR TITLE
Improve Flexmodel and add a snow model

### PR DESCRIPTION
**Pull request by raoulcollenteur**
# Short Description

This PR adds a number of improvements and functionality to the FlexModel recharge model.

- Fix water balance issue for certain combinations of gamma and ks.
- Add a optional snowmelt model according to the degree-day model.
- Make the interception reservoir optional (#314).
- Improved documentation.
- Improved testing with explicit testing of the water balances for each reservoir.
- Improved options to obtain the water balance (fluxes and states)
- Experimental feature for groundwater uptake for testing (may be removed in the future though).

# Checklist before PR can be merged:
- [X] is documented
- [X] PEP8 compliant code
- [X] tests added / passed
- [x] Example Notebook for snow model
- [x] API changes documented in Release Notes

### CODE REVIEW

**Comment by raoulcollenteur** [comment]
@martinvonk can you have a detailed look at this PR? The PR is quite extensive so please take your time and look at the changes carefully. Perhaps also run some tests with your own code.

**Comment by martinvonk** [comment]
> @martinvonk can you have a detailed look at this PR? The PR is quite extensive so please take your time and look at the changes carefully. Perhaps also run some tests with your own code.

Will do. I'll try before Sunday.

**Comment by martinvonk** [comment]
Really cool that this method works!

File: pastas/recharge.py [hunk 1]
```diff
@@ -146,116 +167,347 @@ class FlexModel(RechargeBase):
...
-        parameters.loc[name + "_gamma"] = (4.0, 1e-5, 50.0, True, name)
-        parameters.loc[name + "_simax"] = (2.0, 1e-5, 10.0, False, name)
+        parameters.loc[name + "_ks"] = (100.0, 1e-5, 1e4, True, name)
+        parameters.loc[name + "_gamma"] = (2.0, 1e-5, 20.0, True, name)
         parameters.loc[name + "_kv"] = (1.0, 0.25, 2.0, False, name)
+        if self.interception:
+            parameters.loc[name + "_simax"] = (2.0, 1e-5, 10.0, False, name)
+        if self.snow:
+            parameters.loc[name + "_tp"] = (0.0, -4.0, 4.0, True, name)
+            parameters.loc[name + "_tk"] = (0.0, -4.0, 4.0, True, name)
```
Comment (martinvonk) [inline pastas/recharge.py:221]: Is there an option available to take `tp` and `tk` as one parameter? If I understand correctly `tp` is the snowfall melt parameter and `tk` is a parameter related to the melting point of snow on the ground. I can imagine that these parameters are usually relatively similar. In the publication of _Kavetski and Kuczera_ they also take `tp = tk` for simplicity.

Additionally, what happens when `tp < tk`, then you can have days where is no snow melt but all precipitation falls as rain (and not snow)?

Comment (martinvonk) [reply pastas/recharge.py:221]: On parameter `tp`:

I'm getting really small values for the standard error (e-12).

Comment (raoulcollenteur) [reply pastas/recharge.py:221]: Good points. I just merged these two parameters into one (tt), following Kavetski and Kuczera (2007), the HBV model, and many others studies Commit will follow in a bit. Probably better to start easy and do a proper study where different snow models are tested somewhere in the future.

**Review by martinvonk — APPROVED** [review]
Good job @raoulcollenteur , nice addition.

**Pull request merged**